### DISCUSSION TITLE
How to debug language-server? I don't see console.logs.

### DISCUSSION THREAD

**Discussion by trusktr** [eclipse-langium/langium#1027] [Q&A]
I've put `console.log`s in my validator. No idea where that output is showing when I hit `F5` to test. I don't see anything in `DEBUG CONSOLE`.

**Comment by Lotes** [comment]
The launch config of VSCode has several entries, the first one is for the extension that runs per default on hitting F5. The second is for attaching to the language server. Just run it as the second process and you will be able to debug your validator.

It is always a good idea to test end to end, but I recommend to also add tests, so that you do not have to take this large feedback loop :).

**Reply by trusktr** [reply]
I agree with tests, but we're early stages just trying to get things going.

How exactly do we launch the second part? Might be nice for us to add this into the Getting Started guide.

**Reply by msujew** [reply]
Just as you launch any [debug config in vscode](https://code.visualstudio.com/docs/editor/debugging). The [yeoman generator](https://github.com/langium/langium/blob/0b603a61b4725bc74adf6c21adcff586e129e1e1/packages/generator-langium/langium-template-vscode/.vscode/launch.json#L8-L29) automatically adds the `Attach to Language Server` config to a generated project. As soon as the language server is running, you can attach to it and debug it as you would any other node.js application.

**Comment by msujew** [comment]
In addition to what @Lotes said (which is very helpful for debugging), console logs from the language server appear in the output panel of the vscode instance that runs your extension. You need to select the corresponding output channel first. See:

![image](https://user-images.githubusercontent.com/4377073/232307163-35728512-b002-488b-8a97-91ad07996a19.png)

**Reply by trusktr** [reply]
Nice. Thank you both @Lotes @msujew! I'll try both approaches for a spin, and can recommend some text for the guide.

**Comment by typeholes** [comment]
You probably want tsc watch running or at least to a fresh build before starting to debug.  Out of date source maps are never fun.
